### PR TITLE
Add reverse_continue and step_back commands

### PR DESCRIPTION
**Pull request by Phaiax**
Hi

I wanted to use the [reverse](https://sourceware.org/gdb/wiki/ReverseDebug) [debugging](https://gjbex.github.io/Defensive_programming_and_debugging/BugsAtRuntime/Debuggers/Gdb/gdb_reverse_debugging/) [features](https://sourceware.org/gdb/current/onlinedocs/gdb.html/Process-Record-and-Replay.html) of `gdb`. 

It would be nice if the DAP supports reverse-next but as far as I see that's not supported.

### CODE REVIEW

**Comment by daveleroy** [comment]
I think these need to check `capabilities` to see if the adapter supports `supportsStepBack` and if not issue an error saying the adapter does not support it before issuing the dap command

**Review by daveleroy — COMMENTED** [review]
This seems like a GDB specific thing I don't think suggesting fixes is something we should be doing the adapter should be issuing a relevant error message

File: modules/debugger.py [hunk 1]
```diff
@@ -453,6 +453,12 @@ async def resume(self) -> None:
 		try: await self.active.resume()
 		except core.Error as e: self.console.error(f'Unable to continue: {e}')
 
+	@core.run
+	async def reverse_continue(self) -> None:
+		try: await self.active.reverse_continue()
+		except core.Error as e:
+			self.console.error(f'Unable to reverse continue: Try running the "Debugger: Start recording for reverse debugging" command first. Details: {e}')
```
Comment (daveleroy) [inline modules/debugger.py:460]: This seems like a GDB specific thing I don't think suggesting fixes is something we should be doing the adapter should be issuing a relevant error message

File: modules/dap/session.py [hunk 2]
```diff
@@ -547,6 +559,12 @@ async def step_in(self, granularity: str|None = None):
 	async def step_out(self, granularity: str|None = None):
 		await self.step('stepOut', granularity)
 
+	async def step_back(self, granularity: str|None = None):
+		await self.step('stepBack', granularity)
+
+	async def record(self):
```
Comment (daveleroy) [inline modules/dap/session.py:565]: I think this should live in the gdb adapter its not part of the dap spec

File: modules/commands.py [hunk 3]
```diff
@@ -162,6 +174,12 @@ class Commands:
 		action=lambda debugger: debugger.clear_all_breakpoints(),
 		flags=Command.menu_widget|Command.menu_commands|Command.menu_main,
 	)
+	record = Command (
```
Comment (daveleroy) [inline modules/commands.py:177]: I think this needs to be moved into gdp.py and include the `GDB:` prefix similar to how LLDB commands are done in lldb.py its not part of the dap spec.

File: modules/debugger.py [hunk 4]
```diff
@@ -473,6 +479,17 @@ async def step_out(self) -> None:
...
 		except core.Error as e: self.console.error(f'Unable to step out: {e}')
 
+	@core.run
+	async def step_back(self) -> None:
+		try: await self.active.step_back(granularity=self.stepping_granularity())
+		except core.Error as e:
+			self.console.error(f'Unable to step back: Try running the "Debugger: Start recording for reverse debugging" command first. Details: {e}')
+
+	@core.run
+	async def record(self) -> None:
```
Comment (daveleroy) [inline modules/debugger.py:489]: Same as the `record` in session all this logic should be in the gdb adapter I think

File: modules/debugger.py (hunk 1 above)
Comment (Phaiax) [reply modules/debugger.py]: done, but I found the DAP message a bit confusing:

`Could not step back: Target multi-thread does not support this command. (from exec-step --reverse) - Try running 'target record-full' before stepping back`

File: modules/dap/session.py (hunk 2 above)
Comment (Phaiax) [reply modules/dap/session.py]: done

File: modules/commands.py (hunk 3 above)
Comment (Phaiax) [reply modules/commands.py]: done, not sure if there is a better way to get/call `session.request`

File: modules/debugger.py (hunk 4 above)
Comment (Phaiax) [reply modules/debugger.py]: done

**Comment by Phaiax** [comment]
Hi daveleroy. Just pinging 😸

**Comment by daveleroy** [comment]
> Hi daveleroy. Just pinging 😸

Sorry I should get to this sometimes this weekend

**Comment by daveleroy** [comment]
I made a few changes such as checking that the adapter is gdb before enabling the commands so you should test and make sure its working.

I currently don't have access to gdb but I'll test things when I do

**Pull request merged**